File: libs/moqtail-ts/src/model/data/track_alias.ts
import { TrackAliasError } from '../error'
import { FullTrackName } from './full_track_name'

/**
 * A bidirectional map between track aliases (bigint) and full track names.
 * Used to efficiently look up either by alias or by name, enforcing uniqueness.
 *
 * @public
 */
export class TrackAliasMap {
  private aliasToName = new Map<bigint, FullTrackName>()
  private nameToAlias = new Map<FullTrackName, bigint>()

  /**
   * @public
   * Adds a mapping between an alias and a full track name.
   *
   * @param alias - The alias to associate.
   * @param name - The full track name to associate.
   * @throws :{@link TrackAliasError} if the alias or name is already mapped to a different value.
   */
  addMapping(alias: bigint, name: FullTrackName): void {
    if (this.aliasToName.has(alias)) {
      const existingName = this.aliasToName.get(alias)
      if (existingName === name) return
      throw new TrackAliasError(
        'TrackAliasMap::addMapping(existingName)',
        `Full track name already exists for alias: ${alias}`,
      )
    }
    if (this.nameToAlias.has(name)) {
      const existingAlias = this.nameToAlias.get(name)
      if (existingAlias === alias) return
      throw new TrackAliasError(
        'TrackAliasMap::addMapping(existingAlias)',
        `An alias already exists for full track name`,
      )
    }
    this.aliasToName.set(alias, name)
    this.nameToAlias.set(name, alias)
  }

  /**
   * @public
   * Gets the full track name for a given alias.
   *
   * @param alias - The alias to look up.
   * @returns The associated full track name.
   * @throws :{@link TrackAliasError} if the alias does not exist.
   */
  getNameByAlias(alias: bigint): FullTrackName {
    const name = this.aliasToName.get(alias)
    if (!name) throw new TrackAliasError('TrackAliasMap::getNameByAlias(name)', `Alias: ${alias} doesn't exist`)
    return name
  }

  /**
   * @public
   * Gets the alias for a given full track name.
   *
   * @param name - The full track name to look up.
   * @returns The associated alias.
   * @throws :{@link TrackAliasError} if the name does not exist.
   */
  getAliasByName(name: FullTrackName): bigint {
    const alias = this.nameToAlias.get(name)
    if (alias === undefined) throw new TrackAliasError('TrackAliasMap::getAliasByName(alias)', `Name does not exist`)
    return alias
  }

  /**
   * @public
   * Removes a mapping by alias.
   *
   * @param alias - The alias to remove.
   * @returns The removed full track name, or undefined if not found.
   */
  removeMappingByAlias(alias: bigint): FullTrackName | undefined {
    const name = this.aliasToName.get(alias)
    if (name) {
      this.aliasToName.delete(alias)
      this.nameToAlias.delete(name)
      return name
    }
    return undefined
  }

  /**
   * @public
   * Removes a mapping by full track name.
   *
   * @param name - The full track name to remove.
   * @returns The removed alias, or undefined if not found.
   */
  removeMappingByName(name: FullTrackName): bigint | undefined {
    const alias = this.nameToAlias.get(name)
    if (alias !== undefined) {
      this.nameToAlias.delete(name)
      this.aliasToName.delete(alias)
      return alias
    }
    return undefined
  }

  /**
   * @public
   * Checks if the map contains a given alias.
   *
   * @param alias - The alias to check.
   * @returns True if the alias exists, false otherwise.
   */
  containsAlias(alias: bigint): boolean {
    return this.aliasToName.has(alias)
  }

  /**
   * @public
   * Checks if the map contains a given full track name.
   *
   * @param name - The full track name to check.
   * @returns True if the name exists, false otherwise.
   */
  containsName(name: FullTrackName): boolean {
    return this.nameToAlias.has(name)
  }
}

if (import.meta.vitest) {
  const { describe, test, expect } = import.meta.vitest
  describe('TrackAliasMap', () => {
    test('add and get mapping roundtrip', () => {
      const map = new TrackAliasMap()
      const alias = 42n
      const name = FullTrackName.tryNew('namespace/test', 'bamboozeled')
      map.addMapping(alias, name)
      expect(map.getNameByAlias(alias)).toEqual(name)
      expect(map.getAliasByName(name)).toBe(alias)
    })
    test('add duplicate alias error', () => {
      const map = new TrackAliasMap()
      const alias = 1n
      const name1 = FullTrackName.tryNew('namespace/test', 'bamboozeled')
      const name2 = FullTrackName.tryNew('namespace/test/yeeahboii', 'bamboozeled')
      map.addMapping(alias, name1)
      expect(() => map.addMapping(alias, name2)).toThrow()
    })
  })
}

File: libs/moqtail-rs/src/model/data/track_alias.rs
// Copyright 2025 The MOQtail Authors
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use std::collections::HashMap;

use crate::model::error::ParseError;

use super::full_track_name::FullTrackName;

#[derive(Debug, Default, Clone)]
pub struct TrackAliasMap {
  alias_to_name: HashMap<u64, FullTrackName>,
  name_to_alias: HashMap<FullTrackName, u64>,
}

impl TrackAliasMap {
  pub fn new() -> Self {
    Default::default()
  }

  pub fn add_mapping(&mut self, alias: u64, name: FullTrackName) -> Result<(), ParseError> {
    if let Some(existing_name) = self.alias_to_name.get(&alias) {
      if *existing_name == name {
        return Ok(());
      } else {
        return Err(ParseError::TrackAliasError {
          context: "TrackAliasMap::add_mapping(existing_name)",
          details: format!("Full track name already exists for alias: {alias}"),
        });
      }
    }

    if let Some(existing_alias) = self.name_to_alias.get(&name) {
      if *existing_alias == alias {
        return Ok(());
      } else {
        return Err(ParseError::TrackAliasError {
          context: "TrackAliasMap::add_mapping(existing_alias)",
          details: format!("An alias already exists for full track name: {name:?}"),
        });
      }
    }

    self.alias_to_name.insert(alias, name.clone());
    self.name_to_alias.insert(name, alias);
    Ok(())
  }

  pub fn get_name_by_alias(&self, alias: u64) -> Result<&FullTrackName, ParseError> {
    self
      .alias_to_name
      .get(&alias)
      .ok_or(ParseError::TrackAliasError {
        context: "TrackAliasMap::get_name_by_alias(alias)",
        details: format!("Alias: {alias} doesn't exist"),
      })
  }

  pub fn get_alias_by_name(&self, name: &FullTrackName) -> Result<u64, ParseError> {
    self
      .name_to_alias
      .get(name)
      .copied()
      .ok_or(ParseError::TrackAliasError {
        context: "TrackAliasMap::get_alias_by_name(name)",
        details: format!("Name: {name:?} doesn't exist"),
      })
  }

  pub fn remove_mapping_by_alias(&mut self, alias: u64) -> Option<FullTrackName> {
    if let Some(name) = self.alias_to_name.remove(&alias) {
      self.name_to_alias.remove(&name);
      Some(name)
    } else {
      None
    }
  }

  pub fn remove_mapping_by_name(&mut self, name: &FullTrackName) -> Option<u64> {
    if let Some(alias) = self.name_to_alias.remove(name) {
      self.alias_to_name.remove(&alias);
      Some(alias)
    } else {
      None
    }
  }

  pub fn contains_alias(&self, alias: u64) -> bool {
    self.alias_to_name.contains_key(&alias)
  }

  pub fn contains_name(&self, name: &FullTrackName) -> bool {
    self.name_to_alias.contains_key(name)
  }
}

#[cfg(test)]
mod tests {
  use super::*;

  fn make_name() -> FullTrackName {
    FullTrackName::try_new("namespace/test", "bamboozeled").unwrap()
  }
  fn make_name2() -> FullTrackName {
    FullTrackName::try_new("namespace/test/yeeahboii", "bamboozeled").unwrap()
  }

  #[test]
  fn test_add_and_get_mapping() {
    let mut map = TrackAliasMap::new();
    let alias = 42;
    let name = make_name();
    assert!(map.add_mapping(alias, name.clone()).is_ok());
    assert_eq!(map.get_name_by_alias(alias).unwrap(), &name);
    assert_eq!(map.get_alias_by_name(&name).unwrap(), alias);
  }

  #[test]
  fn test_add_duplicate_alias_error() {
    let mut map = TrackAliasMap::new();
    let alias = 1;
    let name1 = make_name();
    let name2 = make_name2();
    assert!(map.add_mapping(alias, name1.clone()).is_ok());
    let err = map.add_mapping(alias, name2.clone());
    assert!(err.is_err())
  }

  #[test]
  fn test_remove_mapping_by_alias_and_name() {
    let mut map = TrackAliasMap::new();
    let alias = 7;
    let name = make_name();
    map.add_mapping(alias, name.clone()).unwrap();
    assert!(map.remove_mapping_by_alias(alias).is_some());
    assert!(!map.contains_alias(alias));
    assert!(!map.contains_name(&name));
    // Add again and remove by name
    map.add_mapping(alias, name.clone()).unwrap();
    assert!(map.remove_mapping_by_name(&name).is_some());
    assert!(!map.contains_alias(alias));
    assert!(!map.contains_name(&name));
  }

  #[test]
  fn test_get_nonexistent_returns_error() {
    let map = TrackAliasMap::new();
    let alias = 99;
    let name = make_name();
    assert!(map.get_name_by_alias(alias).is_err());
    assert!(map.get_alias_by_name(&name).is_err());
  }
}
